File: src/protocol/node.ts
// 统一节点模型 (一切皆文件) —— 所有本地优先内容收敛为单一命名空间里的可寻址节点。
// 见 docs/design/ai-native-redesign.md §2。纯类型 + 零依赖运行期守卫 (不引 platejs/编辑器):
//   - kind 为主辨识 (不设顶层 mime; file 用 blobRef.mime);
//   - 可辨识联合杀掉 content:unknown 的类型逃逸; note 的 unknown[] 是协议纯度被迫且合理的妥协;
//   - 折叠分步进行 (§3): 步 A 仅 note 物理入库, 其余 kind 类型先就位、随后续折叠落库。
import type { NoteContent } from "./files"
import type { SubscriptionType } from "./subscription"

export type NodeKind = "folder" | "note" | "bookmark" | "file" | "feed" | "thread"

/** 节点引用: 一个节点的稳定句柄 (kind + 不透明 id)。寻址真相 = id; nodePath 是沿 parentId 派生的可变视图。 */
export interface NodeRef {
  kind: NodeKind
  id: string
}

/** Blob 旁存引用 (file 专用): 大二进制存独立 blobs 仓库, 不进节点同步块。 */
export interface BlobRef {
  store: "blobs"
  key: string
  size: number
  mime: string
}

/** 所有 kind 共有的基座 —— 结构上满足 @protocol/sync 的 SyncRecord (id + updatedAt + deletedAt)。 */
export interface BaseNode {
  id: string
  /** 父节点 id (复用 Note 树语义); 合成单根树下 null 只属唯一根。 */
  parentId: string | null
  /** 同级排序键 (fractional index, 见 sort-key.ts)。 */
  sortKey: string
  title: string
  tags: string[]
  createdAt: number
  /** 最后编辑时间戳, 毫秒 (LWW)。 */
  updatedAt: number
  /** 软删除墓碑 (epoch ms); 缺省 = 活跃。删除靠墓碑跨端传播 (见 @protocol/sync)。 */
  deletedAt?: number
  /** kind 专属的额外元数据 (按需)。 */
  meta?: Record<string, unknown>
}

/** 统一节点 —— 按 kind 的可辨识联合。 */
export type Node = BaseNode &
  (
    | { kind: "folder"; content?: null }
    | { kind: "note"; content: NoteContent } // Plate Value (协议不依赖 platejs)
    | { kind: "bookmark"; content: { url: string; description: string; favicon: string } }
    | { kind: "file"; blobRef: BlobRef; content?: null } // Blob 旁存 blobs 仓库, 不进同步
    | {
        kind: "feed"
        content: {
          type: SubscriptionType
          key: string
          favicon: string
          entityLabel?: string
          entityName?: string
          searchKeyword?: string
          searchDomain?: string
        }
      }
    | { kind: "thread"; content: { messages: unknown[] } }
  )

/** 取某 kind 对应的节点子类型, 如 NodeOfKind<"note">。 */
export type NodeOfKind<K extends NodeKind> = Extract<Node, { kind: K }>

/** fs.create 入参 (AI fs.* 写面 §6.1): content 用对应 kind 的 Node content 形态。 */
export type FsCreateInput = {
  kind: NodeKind
  parentId?: string | null
  title?: string
  tags?: string[]
  content?: unknown
}

/** fs.write 补丁 (只改给定字段)。 */
export type FsWritePatch = {
  title?: string
  tags?: string[]
  content?: unknown
  parentId?: string | null
}

/** 笔记节点 —— 折叠步 A 唯一物理入库的 kind。 */
export type NoteNode = NodeOfKind<"note">

/** 全部 NodeKind 的**单一真相源**。tools.ts 的 zod enum、nodes-store 的 ALL_NODE_KINDS 等一律从此派生, 杜绝多份手抄漂移。 */
export const NODE_KINDS: readonly NodeKind[] = [
  "folder",
  "note",
  "bookmark",
  "file",
  "feed",
  "thread",
]

/** 运行期校验 (反序列化标签 params / 深链 / 水合时用)。 */
export function isNodeKind(k: string): k is NodeKind {
  return (NODE_KINDS as readonly string[]).includes(k)
}

/**
 * 隐私净化 (单一函数, fs.list 工具与 fs://nodes 资源同点复用, 防净化漂移; 见设计 §6.3):
 * note 剥正文 content, thread 剥 messages —— 批量列举永不回私密正文/会话 (即便消费方持 fs.notes:read)。
 * 其余 kind (folder/bookmark/file/feed) 的 content 非私密正文, 原样返回。纯函数, 不依赖存储。
 *
 * 用穷尽 switch (无 default) 锁死隐私决策: 新增 NodeKind 而未在此显式分类, 会因缺返回路径
 * 而**编译失败** —— 防「加 kind 忘判私密 → 静默原样放行 → 批量列举泄漏」(红队 §must-fix)。
 */
export function stripNode(n: Node): Node {
  switch (n.kind) {
    case "note":
      return { ...n, content: [] }
    case "thread":
      return { ...n, content: { messages: [] } }
    case "folder":
    case "bookmark":
    case "file":
    case "feed":
      return n
  }
}

File: native/crates/ideall-protocol/src/node.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum NodeKind {
    Folder,
    Note,
    Bookmark,
    File,
    Feed,
    Thread,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BaseNode {
    pub id: String,
    pub parent_id: Option<String>,
    pub sort_key: String,
    pub title: String,
    pub tags: Vec<String>,
    pub created_at: i64,
    pub updated_at: i64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub deleted_at: Option<i64>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub meta: BTreeMap<String, Value>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BlobRef {
    pub store: String,
    pub key: String,
    pub size: u64,
    pub mime: String,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BookmarkContent {
    pub url: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub favicon: String,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SubscriptionType {
    Publisher,
    Entity,
    Tool,
    Search,
    Peer,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FeedContent {
    pub r#type: SubscriptionType,
    pub key: String,
    #[serde(default)]
    pub favicon: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub entity_label: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub entity_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub search_keyword: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub search_domain: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct ThreadContent {
    pub messages: Vec<Value>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum Node {
    Folder {
        #[serde(flatten)]
        base: BaseNode,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        content: Option<()>,
    },
    Note {
        #[serde(flatten)]
        base: BaseNode,
        content: Vec<Value>,
    },
    Bookmark {
        #[serde(flatten)]
        base: BaseNode,
        content: BookmarkContent,
    },
    File {
        #[serde(flatten)]
        base: BaseNode,
        #[serde(rename = "blobRef")]
        blob_ref: BlobRef,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        content: Option<()>,
    },
    Feed {
        #[serde(flatten)]
        base: BaseNode,
        content: FeedContent,
    },
    Thread {
        #[serde(flatten)]
        base: BaseNode,
        content: ThreadContent,
    },
}

impl Node {
    pub fn kind(&self) -> NodeKind {
        match self {
            Self::Folder { .. } => NodeKind::Folder,
            Self::Note { .. } => NodeKind::Note,
            Self::Bookmark { .. } => NodeKind::Bookmark,
            Self::File { .. } => NodeKind::File,
            Self::Feed { .. } => NodeKind::Feed,
            Self::Thread { .. } => NodeKind::Thread,
        }
    }

    pub fn base(&self) -> &BaseNode {
        match self {
            Self::Folder { base, .. }
            | Self::Note { base, .. }
            | Self::Bookmark { base, .. }
            | Self::File { base, .. }
            | Self::Feed { base, .. }
            | Self::Thread { base, .. } => base,
        }
    }

    pub fn base_mut(&mut self) -> &mut BaseNode {
        match self {
            Self::Folder { base, .. }
            | Self::Note { base, .. }
            | Self::Bookmark { base, .. }
            | Self::File { base, .. }
            | Self::Feed { base, .. }
            | Self::Thread { base, .. } => base,
        }
    }

    /// Privacy-preserving projection used for bulk listings.
    pub fn strip_for_listing(&self) -> Self {
        match self {
            Self::Note { base, .. } => Self::Note {
                base: base.clone(),
                content: Vec::new(),
            },
            Self::Thread { base, .. } => Self::Thread {
                base: base.clone(),
                content: ThreadContent {
                    messages: Vec::new(),
                },
            },
            other => other.clone(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> BaseNode {
        BaseNode {
            id: "n1".into(),
            parent_id: None,
            sort_key: "a0".into(),
            title: "测试".into(),
            tags: vec!["private".into()],
            created_at: 10,
            updated_at: 20,
            deleted_at: None,
            meta: BTreeMap::new(),
        }
    }

    #[test]
    fn note_serde_matches_typescript_wire() {
        let raw = serde_json::json!({
            "id": "n1",
            "kind": "note",
            "parentId": null,
            "sortKey": "a0",
            "title": "测试",
            "tags": ["private"],
            "createdAt": 10,
            "updatedAt": 20,
            "content": [{"type": "p", "children": [{"text": "secret"}]}]
        });
        let node: Node = serde_json::from_value(raw.clone()).unwrap();
        assert_eq!(node.kind(), NodeKind::Note);
        assert_eq!(serde_json::to_value(&node).unwrap(), raw);

        let stripped = node.strip_for_listing();
        assert_eq!(
            serde_json::to_value(stripped).unwrap()["content"],
            serde_json::json!([])
        );
    }

    #[test]
    fn thread_listing_removes_messages() {
        let node = Node::Thread {
            base: base(),
            content: ThreadContent {
                messages: vec![serde_json::json!({"role": "user", "content": "secret"})],
            },
        };
        let stripped = serde_json::to_value(node.strip_for_listing()).unwrap();
        assert_eq!(stripped["content"]["messages"], serde_json::json!([]));
    }
}
